File: base/summit/Tune-ILT/Scala/Solutions/Includes/StatFiles.scala
// Databricks notebook source

import org.apache.spark.sql.functions._
import org.apache.spark.sql.DataFrame

def statFileWithPartitions(input_path_para: String , partsCol: List[String],input_type:String = "parquet" ):DataFrame ={
  
    val filter_file_name = if(input_type == "parquet") "name like 'part-%'" else " 1==1"
  
    val  input_path = input_path_para +(input_path_para.takeRight(1) match{
        case "/" => ""
        case _ => "/"
    })//input_path_para has to have / at the end

    val oneMB = 1048576.0 //1MB
    val rawDF =  input_type match {
             case "parquet" => spark.read.parquet(input_path)
             case  "csv"|"tsv" => spark.read.csv(input_path)
    }
    val rawDFStat = rawDF.selectExpr(partsCol:_*).distinct
    val listOfPartitions = rawDFStat.collect().map(r => (partsCol.zip(r.toSeq).toList.map(t => t._1+"="+t._2).mkString("/"),r.toSeq.mkString("_")))

    var statArr = listOfPartitions.map{ case(lp,lv) =>{
        val partPath = lp
        val fullPath = input_path + partPath
        println(fullPath)
        val filesDF = dbutils.fs.ls(fullPath).toDF
        val statDF = filesDF
                     .where(filter_file_name)
                     .agg(count("name").as("num_file")
                         ,(max("size")/oneMB).as("max_size")
                         ,(min("size")/oneMB).as("min_size")
                         ,(avg("size")/oneMB).as("avg_size")
                         ,(sum("size")/oneMB).as("part_size")
                      )
        statDF.withColumn("partition",lit(lv))
    }}
     val statDF = statArr.reduce(_.union(_))
     statDF
}

def statFileFolder(input_path: String ):DataFrame ={
      val oneMB = 1048576.0 //1MB

      val filesDF = dbutils.fs.ls(input_path).toDF
      val statDF = filesDF.where("name like 'part-%'").agg(count("name").as("num_file")
                                                             ,(max("size")/oneMB).as("max_size")
                                                             ,(min("size")/oneMB).as("min_size")
                                                             ,(avg("size")/oneMB).as("avg_size")
                                                             ,(sum("size")/oneMB).as("part_size")
                                                            )
  statDF
}

displayHTML("""<div style="font-size:larger">
  <div>Declared utility methods:</div>
  <li><code>statFileFolder(input_path: String): DataFrame</code></li>
  <li><code>statFileWithPartitions(input_path_para: String, partsCol: List[String], input_type: String): DataFrame</code></li>
</div>""")

File: base/summit/Tune-ILT/Python/Solutions/Includes/StatFiles.py
# Databricks notebook source

from pyspark.sql.functions import * 
from pyspark.sql import DataFrame

def statFileWithPartitions(input_path_para, partsCol, input_type):

  from functools import reduce
  
  # input_path_para has to have / at the end
  if (input_path_para[-1] != "/"):
    input_path_para += "/"

  input_path = input_path_para
  
  oneMB = 1048576.0   #1MB

  if (input_type == "parquet"):
    rawDF = spark.read.parquet(input_path)
  else:
    rawDF = spark.read.csv(input_path)
    
  rawDFStat = rawDF.selectExpr(*partsCol).distinct()  # inject iterable values 
   
  listOfPartitions = rawDFStat.rdd.map(lambda a: ("year=" + str(a[0]) + "/month=" + str(a[1]), str(a[0]) + "_" + str(a[1]))).collect()
      
  def func(args):
    lv = args[1]
    partPath = args[0]
    fullPath = input_path + partPath
    filesDF = spark.createDataFrame(dbutils.fs.ls(fullPath))
    
    if input_type == 'parquet':
      tempDF = filesDF.where("name like 'part-%'")
    else:
      tempDF = filesDF
    
    statDF = (tempDF
      .agg(count("name").alias("num_file")
      ,(max("size")/oneMB).alias("max_size")
      ,(min("size")/oneMB).alias("min_size")
      ,(avg("size")/oneMB).alias("avg_size")
      ,(sum("size")/oneMB).alias("part_size"))
    ) 
    return statDF.withColumn("partition",lit(lv))
  
  statArr = map(func, listOfPartitions)
  statDF = reduce(lambda df1, df2: df1.union(df2), statArr)
  
  return statDF

def statFileFolder(input_path):  
  oneMB = 1048576.0 #1MB
  filesDF = dbutils.fs.ls(input_path).toDF
  statDF = (filesDF
    .where("name like 'part-%'")
    .agg(count("name").alias("num_file")
    ,(max("size")/oneMB).alias("max_size")
    ,(min("size")/oneMB).alias("min_size")
    ,(avg("size")/oneMB).alias("avg_size")
    ,(sum("size")/oneMB).alias("part_size"))
   )
            
  return statDF
